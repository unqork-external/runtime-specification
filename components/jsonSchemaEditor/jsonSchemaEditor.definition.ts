import { Const, Default, DiscriminatorValue, Optional, Property, Required } from '@tsed/schema'
import { JSONSchema7 } from 'json-schema'

import { JsonSchemaEditorStyling } from './jsonSchemaEditor.styling'
import { JsonSchemaEditorTargets } from './jsonSchemaEditorTargets.enum'
import { TrimmedDescription, ViewTargets } from '../../../decorators'
import { BaseComponentDefinition } from '../../base-component-interface'
import { StandardArrayNestable } from '../../nestables'
import type { SignalTargets } from '../../signals'

@DiscriminatorValue('jsonSchemaEditor')
@ViewTargets(JsonSchemaEditorTargets)
export class JsonSchemaEditorComponentDefinition extends BaseComponentDefinition {
  @Const('jsonSchemaEditor')
  type: 'jsonSchemaEditor' = 'jsonSchemaEditor' as const

  @Required()
  @TrimmedDescription(`
  The schema you wish to render. Must be valid JSON Schema 7
  `)
  schema: JSONSchema7

  @Optional()
  @Default(false)
  @TrimmedDescription(`
  Set to true to use the form library instead of Unqork components
  `)
  useRawForm = false

  @Optional()
  @Default(false)
  @TrimmedDescription(`
    Fallback to raw form if an error occurs while rendering the Unqork components
  `)
  fallbackToRawForm = false

  @Optional()
  declare styling?: JsonSchemaEditorStyling

  @Optional()
  declare signals?: SignalTargets<JsonSchemaEditorTargets>

  @Optional()
  nestables = {
    childIds: new StandardArrayNestable(),
  }

  @Property()
  childIds: string[] = []
}
