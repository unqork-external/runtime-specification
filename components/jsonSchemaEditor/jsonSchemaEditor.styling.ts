import { Description, Property } from '@tsed/schema'

import { JsonSchemaEditorTargets } from './jsonSchemaEditorTargets.enum'
import { Css } from '../../../decorators/schema/css.decorator'
import { type CssClassOrProps } from '../../styling/cssClassOrProps.type'
import { StylingModel } from '../../styling/styling.model'
import { TargetsModel } from '../../viewTargets/targets.model'

export class JsonSchemaEditorStyleTargets implements TargetsModel<JsonSchemaEditorTargets> {
  @Css()
  @Description('Adds styles to jsonSchemaEditor root ComponentWrapper.')
  [JsonSchemaEditorTargets.root]: CssClassOrProps
}

export class JsonSchemaEditorStyling extends StylingModel<JsonSchemaEditorStyleTargets> {
  @Property()
  declare targets: JsonSchemaEditorStyleTargets
}
