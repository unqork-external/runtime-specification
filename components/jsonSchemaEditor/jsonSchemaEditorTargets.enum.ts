export enum JsonSchemaEditorTargets {
  root = 'root',
}
