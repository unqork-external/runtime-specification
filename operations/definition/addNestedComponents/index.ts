export * from './addNestedComponents.model'
export * from './addNestedComponents.options'
