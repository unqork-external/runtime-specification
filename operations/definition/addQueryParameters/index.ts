export * from './addQueryParameters.model'
export * from './addQueryParameters.options'
