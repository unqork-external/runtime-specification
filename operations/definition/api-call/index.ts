export * from './apiCall.type'
export * from './method.type'
export * from './serviceType.type'
export * from './apiCall.model'
export * from './apiCall.options'
