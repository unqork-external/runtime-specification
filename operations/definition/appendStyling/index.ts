export * from './appendStyling.model'
export * from './appendStyling.options'
