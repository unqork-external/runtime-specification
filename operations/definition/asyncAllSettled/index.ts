export * from './asyncAllSettled.model'
export * from './asyncAllSettled.options'
