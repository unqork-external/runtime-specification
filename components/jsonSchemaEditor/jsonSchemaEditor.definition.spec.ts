import { JsonSchemaEditorComponentDefinition } from './jsonSchemaEditor.definition'
import { generateSchemaAndValidate } from '../../../utilities'

describe('JsonSchemaEditor Contract Tests', function () {
  it('should ensure the contract is clear', function () {
    const { schema } = generateSchemaAndValidate(JsonSchemaEditorComponentDefinition)

    expect(schema.required).toMatchSnapshot()
    expect(schema.properties).toMatchSnapshot()
  })
})
