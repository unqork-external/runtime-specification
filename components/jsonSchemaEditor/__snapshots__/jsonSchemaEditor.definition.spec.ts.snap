// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JsonSchemaEditor Contract Tests should ensure the contract is clear 1`] = `
[
  "key",
  "type",
  "schema",
]
`;

exports[`JsonSchemaEditor Contract Tests should ensure the contract is clear 2`] = `
{
  "childIds": {
    "type": "array",
  },
  "executable": {
    "default": true,
    "description": "Defines whether or not the object will respond to events or fire operations. When \`true\`, the object will perform and behave as defined in its spec. When \`false\`, the object will no longer be responsive to the system.",
    "type": "boolean",
  },
  "fallbackToRawForm": {
    "default": false,
    "description": "Fallback to raw form if an error occurs while rendering the Unqork components",
    "type": "object",
  },
  "inputs": {
    "items": {
      "$ref": "#/definitions/InputRef",
    },
    "type": "array",
  },
  "key": {
    "description": "'key' is the user-defined, unique identifier for an object within a module. It doesn't have to be unique across different modules. Objects within the runtime use keys to target other objects.",
    "examples": [
      "I am the key",
      "bestKeyEver",
    ],
    "minLength": 1,
    "type": "string",
  },
  "nestables": {
    "type": "object",
  },
  "package": {
    "$ref": "#/definitions/PackageModel",
  },
  "schema": {
    "description": "The schema you wish to render. Must be valid JSON Schema 7",
    "type": "object",
  },
  "signals": {
    "description": "An object which contains all of the signals for the component. The signals object can have any number of target keys as defined in the component definition which each contain an array of Signals.",
    "examples": [
      {
        "args": {
          "inputs": {
            "propertyPath": "value",
            "required": true,
            "targetKey": "firstName",
          },
        },
        "id": "firstNameWatcher",
        "operations": [
          {
            "options": {
              "logType": "warn",
              "thingToPrint": "firstNameWatcher operations fired: value watcher triggered",
            },
            "type": "PRINTER",
          },
        ],
        "type": "onWatch",
      },
      {
        "id": "textfieldOnBlur",
        "operations": [
          {
            "options": {
              "logType": "warn",
              "thingToPrint": "textfield received blur event from DOM",
            },
            "type": "PRINTER",
          },
        ],
        "type": "onBlur",
      },
      "<Example> Prompt: "I want a textfield that will trigger the root execute signal on myValidationComponent when the textfield input receives a blur event." Response: [{"key":"yourGeneratedTextfield","type":"textfield","signals":{"input":[{"id":"textfieldOnBlur","operations":[{"type":"EXECUTE","options":{"targetKey":"myValidationComponent","signalTarget":"root","shouldExecute":true}}],"type":"onBlur"}]}}] </Example>",
    ],
    "type": "object",
  },
  "styling": {
    "$ref": "#/definitions/JsonSchemaEditorStyling",
  },
  "type": {
    "const": "jsonSchemaEditor",
    "description": "Name of the type of component.",
    "examples": [
      "jsonSchemaEditor",
    ],
    "minLength": 1,
    "type": "string",
  },
  "useRawForm": {
    "default": false,
    "description": "Set to true to use the form library instead of Unqork components",
    "type": "object",
  },
  "validation": {
    "$ref": "#/definitions/Validation",
  },
  "value": {
    "description": "Current value of the component",
    "type": [
      "null",
      "integer",
      "number",
      "string",
      "boolean",
      "array",
      "object",
    ],
  },
}
`;
